### PR TITLE
FAISS and embedding support 

### PR DESCRIPTION
**Pull request by hwchase17**
also adds embeddings and an in memory docstore

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -12,5 +12,6 @@ google-search-results
 playwright
 wikipedia
 huggingface_hub
+faiss
```
Comment (sjwhitmore) [inline requirements.txt:15]: did you mean for this to be faiss-cpu? faiss itself doesnt show up as a package for me, looks like you can pip install faiss-cpu or faiss-gpu (From README here --https://github.com/facebookresearch/faiss)

**Comment by sjwhitmore** [comment]
trying to test this locally & getting a weird error around openai import, have you seen this before? in both jupyter notebook & vanilla python shell, I get an error about openai import from pydantic, but can run import openai myself with no issues.
image

**Comment by sjwhitmore** [comment]
aha! found the issue in the root_validator for the OpenAIEmbeddings class, I think you need to add matplotlib explicitly to requirements.txt
image

**Comment by sjwhitmore** [comment]
ok, i finally got it to work by adding
matplotlib
plotly
scipy 
sklearn
explicitly to requirements.txt

File: langchain/embeddings/openai.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                " `OPENAI_API_KEY` which contains it."
+            )
+        try:
+            from openai.embeddings_utils import get_embedding
+
+            values["embedding_func"] = get_embedding
+        except ImportError:
+            raise ValueError(
+                "Could not import openai python package. "
+                "Please it install it with `pip install openai`."
```
Comment (sjwhitmore) [inline langchain/embeddings/openai.py:37]: This error message was a bit misleading in the end cause I had openai installed & was missing the dependencies specifically for the embedding_utils import... wonder if changing the text here to be more specific about the import to test will help if others run into this problem?

**Comment by sjwhitmore** [comment]
one comment about the error message that threw me... otherwise this is great I want to use it right away!!

File: requirements.txt (hunk 1 above)
Comment (sjwhitmore) [reply requirements.txt:15]: also you will need to add the following to the requirements (consolidating from chain above):
matplotlib
plotly
scipy
sklearn

**Comment by hwchase17** [comment]
> trying to test this locally & getting a weird error around openai import, have you seen this before? in both jupyter notebook & vanilla python shell, I get an error about openai import from pydantic, but can run import openai myself with no issues. image

hmmm i have not seen

**Pull request merged**